keps/sig-node/2008-forensic-container-checkpointing/README.md

```diff
@@ -0,0 +1,256 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Provide an interface to trigger a container checkpoint for forensic analysis.
+
+## Motivation
+
+Container checkpointing provides the functionality to take a snapshot of a
+running container. The checkpointed container can be transferred to another
+node and the original container will never know that it was checkpointed.
+
+Restoring the container in a sandboxed environment provides a mean to
+forensically analyse a copy of the container to understand if it might
+have been a possible threat. As the analysis is happening on a copy of
+the original container a possible attacker of the original container
+will not be aware of any sandboxed analysis.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the CRI API.
+This includes extending the *kubelet* API to support checkpointing single
+containers with the forensic use case in mind.
+
+### Non-Goals
+
+Although *checkpoint* and *restore* can be used to implement container
+migration this KEP is only about enabling the forensic use case. Checkpointing
+a pod is not part of this proposal and left for future enhancements.
+
+## Proposal
+
+### Implementation
+
+For the forensic use case we want to offer the functionality to checkpoint a
+container out of a running Pod without stopping the checkpointed container or
+letting the container know that it was checkpointed.
+
+The corresponding code changes for the forensic use case can be found in the
+following pull request:
+
+* https://github.com/kubernetes/kubernetes/pull/104907
+
+The goal is to introduce *checkpoint* and *restore* in a bottom-up approach.
+In a first step we only want to extend the CRI API to trigger a checkpoint
+by the container engine and to have the low level primitives in the *kubelet*
+to trigger a checkpoint. It is necessary to enable the feature gate
+`ContainerCheckpointRestore` to be able to checkpoint containers.
+
+In the corresponding pull request a checkpoint is triggered using the *kubelet*
+API:
+
+```
+curl -skv -X POST "https://localhost:10250/checkpoint/default/counters/wildfly"
```

Comment:
This should be seen as a highly privileged operation. How will it be authorized?

This seems on par with exec permissions, but exec uses the highly overloaded `proxy` subresource for the authorization check. I'd prefer to use a new dedicated subresource, preferably scoped to the container for this. I'm not aware of us doing this anywhere else, but it would be interesting if it required 2 authorization checks: that you have access to the checkpoint subresource on the _node_ (since it's hitting a node API directly), and the checkpoint subresource on the _pod_.

Reply:
> This should be seen as a highly privileged operation. How will it be authorized?

Not sure I understand this question correctly. To do this operation the normal kubelet API access restrictions apply. I can only do it without authorization if I set `KUBELET_AUTHORIZATION_WEBHOOK="false"` and `KUBELET_FLAGS="--anonymous-auth=true --authorization-mode=AlwaysAllow"`. Isn't that enough? Are you thinking of additional checks?

Reply:
Kubelet authorization (without `AlwaysAllow`) works by sending a `SubjectAccessReview` request to the apiserver, which includes (among other things) the resource being checked (currently always node) and a subresource. You can find the mapping for the existing Kubelet API here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/auth.go#L59-L112

At the bottom of that link, you'll see that we use a dedicated subresource for some requests (logs, stats & metrics) but everything else just uses the highly overloaded `proxy` subresource.

I'd like for all new Kubelet APIs to use a new dedicated subresource (e.g. `checkpoint`). My comment about checking the subresource against the node & pod probably doesn't make sense as a 1-off (maybe something to consider if we ever overhaul the Kubelet API)